unittests/Interpreter/InterpreterInfra.h

```diff
@@ -0,0 +1,153 @@
+
+#include "llvm/Support/KnownBits.h"
+#include "llvm/IR/ConstantRange.h"
+
+#include "souper/Infer/Interpreter.h"
+#include "souper/Infer/AbstractInterpreter.h"
+
+#include <bitset>
+
+// width used for transfer function tests
+constexpr int WIDTH = 4;
+
+namespace souper {
+
+    struct EvalValueKB : public EvalValue {
+      llvm::KnownBits ValueKB;
+
+      EvalValueKB(llvm::KnownBits Val);
+      EvalValueKB(EvalValue &Val);
+      llvm::KnownBits getValueKB();
+    };
+
+    class KBTesting {
+      // wrapper over @llvm::KnownBitsAnalysis::mergeKnownBits()
+      EvalValueKB merge(EvalValueKB a, EvalValueKB b);
+
+      llvm::KnownBits setLowest(llvm::KnownBits x);
+      llvm::KnownBits clearLowest(llvm::KnownBits x);
+
+      EvalValueKB bruteForce(llvm::KnownBits x, llvm::KnownBits y, Inst::Kind Pred);
+    public:
+      static bool nextKB(llvm::KnownBits &x);
+      bool testFn(Inst::Kind pred);
+    };
+
+
+
+    class CRTesting {
+      bool ok(const llvm::ConstantRange &R, const bool Table[]);
+
+      // Find the largest hole and build a llvm::ConstantRange around it
+      llvm::ConstantRange bestCR(const bool Table[], const int Width);
+      llvm::ConstantRange exhaustive(const llvm::ConstantRange &L, const llvm::ConstantRange &R,
+			       Inst::Kind pred, const llvm::ConstantRange &Untrusted);
+
+      void check(const llvm::ConstantRange &L, const llvm::ConstantRange &R, Inst::Kind pred,
+		 double &FastBits, double &PreciseBits, int &Count, int &PreciseCount);
+
+    public:
+
+      static llvm::ConstantRange nextCR(const llvm::ConstantRange &CR);
+      bool testFn(Inst::Kind pred);
+    };
+
+ namespace TestingUtil {
+
+    void exhaustiveKBCRReduction(llvm::KnownBits &KB, llvm::ConstantRange &CR);
+
+    template<int SetSize>
+    std::bitset<SetSize> concretizeKB(const llvm::KnownBits &KB) {
+      assert(SetSize == (1 << KB.getBitWidth()));
+
+      std::bitset<SetSize> ResultSet(1);
+      for (int Cnt = 0; Cnt < SetSize; Cnt++) {
+
+	llvm::KnownBits ValueKB;
+	auto Tmp = llvm::APInt(KB.getBitWidth(), Cnt);
+	ValueKB.One = Tmp;
+	ValueKB.Zero = ~Tmp;
+
+	if (KnownBitsAnalysis::isConflictingKB(ValueKB, KB))
+	  ResultSet[Cnt] = 0; // exclude this guy
+      }
+
+      return ResultSet;
+    }
+
+    template<int SetSize>
+    std::bitset<SetSize> concretizeCR(const llvm::ConstantRange &CR) {
+      assert(SetSize == (1 << CR.getBitWidth()));
+
+      std::bitset<SetSize> ResultSet(0);
+      if (CR.isFullSet()) {
+	ResultSet.set();
+	return ResultSet;
+      }
+
+      for (int Cnt = 0; Cnt < SetSize; Cnt++) {
+	if (CR.contains(llvm::APInt(WIDTH, Cnt)))
+	  ResultSet[Cnt] = 1; // include this guy
+      }
+      return ResultSet;
+    }
+
+    template<int SetSize>
+    llvm::ConstantRange abstractizeCR(std::bitset<SetSize> &ResultSet) {
+      llvm::ConstantRange ResultCR(WIDTH, false);
+      for (unsigned I = 0; I < SetSize; I++) {
+	if (ResultSet[I]) {
+	  ResultCR = ResultCR.unionWith(llvm::APInt(WIDTH, I));
```

Comment:
I don't think this is going to give the most precise result, can you look back at my code and see if there's a better way to do this? I believe I had a different approach

Reply:
the problem is ConstantRange isn't a lattice and there's no unique LUB of arbitrary sets of elements

Reply:
>  can you look back at my code and see if there's a better way to do this? I believe I had a different approach

I am not aware of any code you wrote related to abstractizing ConstantRanges. Your  code that I copied to this PR was just related to  CR KB reduction.

Reply:
https://github.com/regehr/llvm-dataflow-research/blob/master/test-xfer.cpp#L74

Reply:
So, this thing (the link pasted) that you are talking about is already taken care of in the `exhaustiveCR` function. See the call to `bestCR` in that function. 

The quoted code you have commented against is just abstractizing the ResultSet for CR. I believe the LLVM's `unionWith` function, which looks quite elaborate and correct, to be enough for this purpose. Are you saying that LLVM's `unionWith` function won't give the most precise constant range given set of numbers?

Reply:
I'm telling you that there is no most precise constant range